### PR TITLE
feat: Detects issues with downloading images and starting containers

### PR DESCRIPTION
**Pull request by tolusha**
Signed-off-by: Anatolii Bazko <abazko@redhat.com>

### What does this PR do?
PR adds check on downloading and staring images to fail immediately and don't wait 2m timeout

### Screenshot/screencast of this PR
❯ Operator pod bootstrap
      ✔ Scheduling...done.
      ✖ Downloading images...failed
        → Failed to download image: rpc error: code = Unknown desc = Error response from daemon: pull access denied for test, repository does not exist or may require 'do

❯ Operator pod bootstrap
      ✔ Scheduling...done.
      ✔ Downloading images...done.
      ✖ Starting...failed
        → Failed to start container(s): back-off 10s restarting failed container=che-operator pod=che-operator-756f87dbc-z5knk_che(fe6e4cc9-1349-4a0a-85b7-0a69f56fcde6)

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/17942
https://github.com/eclipse/che/issues/17937

### How to test this PR?
1. Build sources
2. `bin/run server:start --platform minikube  --che-operator-image test:test` to see the error message

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

**Review by sleshchenko — APPROVED** [review]
Looks great!

**Review by AndrienkoAleksandr — APPROVED** [review]
Tested with openshift:

```
./run server:start --platform minikube --che-operator-image test:test
```

Works fine.

**Comment by AndrienkoAleksandr** [comment]
@tolusha Fix import order, please.

File: src/api/kube.ts [hunk 1]
```diff
@@ -348,7 +348,7 @@ export class KubeHelper {
     }
   }
 
-  async getPodListByLabel(namespace= '', labelSelector: string): Promise<V1Pod[]> {
+  async getPodListByLabel(namespace = '', labelSelector: string): Promise<V1Pod[]> {
```
Comment (mmorhun) [inline src/api/kube.ts:351]: Please remove default empty value

File: src/api/kube.ts [hunk 2]
```diff
@@ -659,24 +659,26 @@ export class KubeHelper {
...
+    const pods = await this.getPodListByLabel(namespace, selector)
+    if (!pods.length) {
       return
     }
 
-    if (!res.body.items[0].status || !res.body.items[0].status.phase) {
-      return
+    for (const pod of pods) {
+      if (pod.status && pod.status.phase === desiredPhase) {
+        if (pod.status && pod.status.containerStatuses) {
```
Comment (mmorhun) [inline src/api/kube.ts:673]: I think this two ifs could be joined

File: src/tasks/kube.ts [hunk 3]
```diff
@@ -7,50 +7,76 @@
...
-            phase = await this.kube.getPodPhase(selector, namespace)
-          } catch (err) {
-            // not able to grab current phase
-            this.debug(err)
-          }
-          // wait only if not yet running
-          if (phase !== 'Running') {
-            await this.kube.waitForPodPending(selector, namespace)
+          // tslint:disable-next-line: prefer-while
+          for (; ;) {
```
Comment (mmorhun) [inline src/tasks/kube.ts:27]: Maybe it is still better to prefer while? The same all such loops in the PR.

File: src/tasks/kube.ts [hunk 4]
```diff
@@ -7,50 +7,76 @@
...
-  kube: KubeHelper
-  debug = require('debug')
+  kubeHelper: KubeHelper
   constructor(flags?: any) {
-    this.kube = new KubeHelper(flags)
+    this.kubeHelper = new KubeHelper(flags)
   }
 
-  podStartTasks(_command: Command, selector: string, namespace = ''): Listr {
+  podStartTasks(selector: string, namespace = ''): Listr {
```
Comment (mmorhun) [inline src/tasks/kube.ts:21]: Please remove empty string default value

**Comment by tolusha** [comment]
@mmorhun 
Remarks are fixed.

**Comment by AndrienkoAleksandr** [comment]
[retest]

**Review by mmorhun — APPROVED** [review]

**Pull request merged**